// otp-verification/esm2015/lib/otp-verification.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, ViewChild, EventEmitter, Input, ElementRef } from '@angular/core';
import { OtpVerificationService } from './otp-verification.service';
export class OtpVerificationComponent {
    /**
     * @param {?} otpVerificationService
     */
    constructor(otpVerificationService) {
        this.otpVerificationService = otpVerificationService;
        this.otpOut = new EventEmitter();
        this.specialKeys = [
            'CapsLock', 'ControlLeft', 'ControlRight', 'ShiftLeft', 'ShiftRight',
            'AltLeft', 'AltRight', 'Escape', 'Backquote', 'ArrowUp',
            'Space', 'ArrowDown', 'Delete', 'Insert', 'End',
            'Home', 'PageDown', 'ArrowRight', 'WakeUp'
        ];
        this.inputLimt = [];
    }
    /**
     * @param {?} event
     * @param {?} next
     * @param {?} previous
     * @param {?} current
     * @return {?}
     */
    nextDigit(event, next, previous, current) {
        if (this.specialKeys.includes(event.code)) {
            return;
        }
        if (!this.otpVerificationService.isBackspace(event.key) && !this.isAlphanumeric && isNaN(event.key)) {
            current.value = '';
            return;
        }
        if (this.otpVerificationService.isBackspace(event.key)) {
            if (current.value !== '' && previous) {
                current.value = '';
                previous.focus();
            }
            else if (previous) {
                previous.value = '';
                previous.focus();
            }
            else {
                current.value = '';
            }
        }
        else if (!this.otpVerificationService.isBackspace(event.key)) {
            if (event.key.length === 1) {
                if (event.getModifierState('CapsLock')) {
                    current.value = event.key.toUpperCase();
                }
                else {
                    current.value = event.key;
                }
            }
            setTimeout((/**
             * @return {?}
             */
            () => {
                if (next) {
                    next.focus();
                }
            }));
        }
        this.returnOtp();
    }
    /**
     * @return {?}
     */
    returnOtp() {
        /** @type {?} */
        let otp = '';
        Array.from(this.otpVal.nativeElement.children).forEach((/**
         * @param {?} child
         * @return {?}
         */
        child => {
            if (!child['value']) {
                return;
            }
            otp += child['value'];
        }));
        if (otp.length === 4) {
            this.otpOut.emit(otp);
        }
    }
}
OtpVerificationComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-otp',
                template: "<div class=\"angular-otp__wrap\">\n  <form autocomplete=\"off\" novalidate>\n    <div class=\"angular-otp__input-div\" #otp>\n      <input type=\"text\" angularOnlyNumber [isAlphanumeric]=\"isAlphanumeric\" #otpDigit1 autofocus\n        (keydown)=\"nextDigit($event, otpDigit2, 0, otpDigit1)\" name=\"digit1\" placeholder=\"-\" maxlength=\"1\">\n      <input type=\"text\" angularOnlyNumber [isAlphanumeric]=\"isAlphanumeric\" #otpDigit2\n        (keydown)=\"nextDigit($event, otpDigit3, otpDigit1, otpDigit2)\" name=\"digit2\" placeholder=\"-\" maxlength=\"1\">\n      <input type=\"text\" angularOnlyNumber [isAlphanumeric]=\"isAlphanumeric\" #otpDigit3\n        (keydown)=\"nextDigit($event, otpDigit4, otpDigit2, otpDigit3)\" name=\"digit3\" placeholder=\"-\" maxlength=\"1\">\n      <input type=\"text\" angularOnlyNumber [isAlphanumeric]=\"isAlphanumeric\" #otpDigit4\n        (keydown)=\"nextDigit($event, 0, otpDigit3, otpDigit4)\" name=\"digit4\" placeholder=\"-\" maxlength=\"1\">\n    </div>\n  </form>\n</div>",
                styles: [".angular-otp__input-div{border-bottom:2px solid #d7d7d7;justify-content:center;width:-webkit-fit-content;width:-moz-fit-content;width:fit-content}.angular-otp__input-div input{width:42px;height:45px;background:0 0;border:none;text-align:center;font-size:25px;font-weight:600;outline:0}"]
            }] }
];
/** @nocollapse */
OtpVerificationComponent.ctorParameters = () => [
    { type: OtpVerificationService }
];
OtpVerificationComponent.propDecorators = {
    isAlphanumeric: [{ type: Input }],
    otpOut: [{ type: Output }],
    otpVal: [{ type: ViewChild, args: ['otp',] }]
};
if (false) {
    /** @type {?} */
    OtpVerificationComponent.prototype.isAlphanumeric;
    /** @type {?} */
    OtpVerificationComponent.prototype.otpOut;
    /** @type {?} */
    OtpVerificationComponent.prototype.otpVal;
    /** @type {?} */
    OtpVerificationComponent.prototype.specialKeys;
    /** @type {?} */
    OtpVerificationComponent.prototype.inputLimt;
    /**
     * @type {?}
     * @private
     */
    OtpVerificationComponent.prototype.otpVerificationService;
}
